_posts/2021-12-09-kafka-segment-retention.md

```diff
@@ -0,0 +1,259 @@
+Dumping /tmp/kafka-logs-0/__strimzi_canary-0/00000000000000000000.timeindex
+timestamp: 1638100314372 offset: 28
+timestamp: 1638100454372 offset: 56
+timestamp: 1638100594371 offset: 84
+```
+
+## Let's talk about rolling segments
+
+A new segment is rolled when some specific conditions are met.
+One condition is when the maximum segment size is reached, as specified by the configuration parameter `log.segment.bytes` (1 GiB by default).
+Another condition is based on the `log.roll.ms` or `log.roll.hours` (7 days by default) parameters. In this case, the segment is rolled when the configured time since the producer timestamp of the first record in the segment (or since the creation time if there is no timestamp) has elapsed.
+The first condition that is met will trigger the rolling of a new segment.
+It is also worth noticing that, as a not so usual use case, the producer timestamps in the records could be not ordered, so not having the older record as the first one, due to retries or the specific business logic of the producer.
+
+Another useful parameter is `log.roll.jitter.ms` which sets a maximum jitter when it's time to roll a segment,. Jitters are time intervals to avoid many segments being rolled at same time causing high contention on the disk.
+A jitter is randomly generated for each segment.
+
+The above conditions, by size or time, are the well known ones but not everyone knows that there is a third condition.
+
+A segment is rolled even when the corresponding index (or timeindex) is full.
+The index and timeindex share the same maximum size which is defined by the `log.index.size.max.bytes` configuration parameter and it is 10 MB by default.
+Let's consider the default log segment maximum size, which is 1 GiB.
+We know that because `log.index.interval.bytes` is 4096 bytes by default, an entry is added in the index every 4096 bytes of records.
+It means that for a 1 GiB segment size, 1 GiB / 4096 bytes = 262144 entries are added to the index. This equals 2 MB of index (262144 * 8 bytes).
+The default index size of 10 MB is enough to handle a segment size of 5 GiB.
+This means that reducing the index size or increasing the segment size will mean a new segment is rolled when the index is full and not when the requested segment size is reached.
+So, setting a segment size larger than 5 GiB without increasing the index size is useless, because the broker will roll a new segment when the index is full.
+
+By increasing the segment size over 5 GiB, you would also need to increase the index file size as well. Likewise, if you decide to reduce the index file size, you should probably decrease the segment size as well.
+
+The timeindex might need attention as well. Because each timeindex entry is 1.5x bigger than an entry in the index (12 bytes versus 8 bytes), it can fill up earlier and cause a new segment to be rolled.
```

Comment:
What do you mean by `need attention`? What should a user do? Especially as both indexes share the same size setting.

Reply:
It's a poor way of writing what I meant for sure :-)
There is no needed intervention from a user pov. It was a kind of introduction to the next sentence ..."Because each ...." so raising the point that you should not focus just on index file but even on timeindex because it can be filled earlier.
So it needs "attention" :-D